File: app/src/main/java/com/ywwynm/everythingdone/adapters/MultiChoiceAdapter.java
package com.ywwynm.everythingdone.adapters;

import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.List;

/**
 * Created by ywwynm on 2015/9/16.
 * Adapter to offer multi select for items.
 */
public abstract class MultiChoiceAdapter extends RecyclerView.Adapter<BaseViewHolder> {

    public static final String TAG = "MultiChoiceAdapter";

    protected boolean[] mPicked;

    @Override
    public abstract BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType);

    @Override
    public abstract void onBindViewHolder(BaseViewHolder viewHolder, int position);

    @Override
    public abstract int getItemCount();

    public void togglePick(int position) {
        mPicked[position] = !mPicked[position];
        notifyItemChanged(position);
    }

    public void pick(int position) {
        mPicked[position] = true;
    }

    public void pickAll() {
        for (int i = 0; i < mPicked.length; i++) {
            mPicked[i] = true;
        }
    }

    public void unpickAll() {
        for (int i = 0; i < mPicked.length; i++) {
            mPicked[i] = false;
        }
    }

    public void pick(List<Integer> positions) {
        for (Integer position : positions) {
            mPicked[position] = true;
        }
    }
}

File: app/src/main/java/com/ywwynm/everythingdone/adapters/MultiChoiceAdapter.kt
package com.ywwynm.everythingdone.adapters

import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup

/**
 * Created by ywwynm on 2015/9/16.
 * Translated to Kotlin by ywwynm and Claude Opus 4.7 on 2026/5/20.
 * Adapter to offer multi select for items.
 */
abstract class MultiChoiceAdapter : RecyclerView.Adapter<BaseViewHolder>() {

    @JvmField
    protected var mPicked: BooleanArray? = null

    abstract override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder

    abstract override fun onBindViewHolder(viewHolder: BaseViewHolder, position: Int)

    abstract override fun getItemCount(): Int

    open fun togglePick(position: Int) {
        mPicked!![position] = !mPicked!![position]
        notifyItemChanged(position)
    }

    open fun pick(position: Int) {
        mPicked!![position] = true
    }

    open fun pickAll() {
        for (i in mPicked!!.indices) {
            mPicked!![i] = true
        }
    }

    open fun unpickAll() {
        for (i in mPicked!!.indices) {
            mPicked!![i] = false
        }
    }

    open fun pick(positions: List<Int?>?) {
        for (position in positions!!) {
            mPicked!![position!!] = true
        }
    }

    companion object {
        const val TAG: String = "MultiChoiceAdapter"
    }
}
